### ISSUE TITLE
Async-safe context managers

### ISSUE THREAD

**Issue by felix-hilden** [felix-hilden/tekore#242]
Currently our context managers are not async safe. So let's make them safe or *at least* document the fact that they are not safe.

**Comment by felix-hilden** [comment]
Locking would have the disadvantage of slowing async down, which `contextvars` seems to avoid by simply providing a new context each time. So we could use that instead of a lock.

**Comment by felix-hilden** [comment]
But we still have some designing to do. The context vars are async safe, but what should happen when a token is directly set? I think something like this:

- If there are no contexts: setting the token modifies the underlying, persistent token value
- If there is a local context: setting the token modifies the context value too
- If there is a context elsewhere: setting the token must not interfere with that context

So in a nutshell 1) we would guarantee that contexts are safe from modifications elsewhere and 2) setting a token works as expected locally and 3) there are no guarantees with simply setting tokens.

Thoughts?

**Comment by felix-hilden** [comment]
Then again, we should probably let setting the token inside a context be thrown away when the context exits, like it is done currently. Yeah, so that only without any local contexts setting a token modifies the true persistent value.

**Comment by felix-hilden** [comment]
Sadly the `contextvars` backport for Python 3.6 does not deal with async, so we'll have to simply document the fact that the context managers are async safe only on 3.7+
